### ISSUE TITLE
Errors injecting multi cluster configuration on Service Fabric

### ISSUE THREAD

**Issue by charleszipp** [dotnet/orleans#3174]
I am attempting to setup a POC to demonstrate an orleans multi-cluster setup for geo-replication. When i startup either cluster i am getting the following error. We are looking at orleans primarily for geo-replication its critical we are able to prove that it can do this inside service fabric.

I am using 1.5.0-beta1 version of all the orleans binaries.

```
Audit.Engagements.Actors.exe Warning: 0 : [2017-06-26 13:19:42.591 GMT    16	WARNING	103905	MultiClusterOracle	10.0.0.4:0]	-3 Publish to silo  (scus) failed	
Exc level 0: System.ArgumentNullException: Value cannot be null.
Parameter name: key
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at Orleans.GrainFactory.GetSystemTarget[TGrainInterface](GrainId grainId, SiloAddress destination)
   at Orleans.Runtime.MultiClusterNetwork.MultiClusterOracle.SiloGossipWorker.<Publish>d__9.MoveNext()
```
Subsequently i get this error every minute or so.

```
Audit.Engagements.Actors.exe Warning: 0 : [2017-06-26 13:20:10.084 GMT    17	WARNING	103905	MultiClusterOracle	10.0.0.4:0]	Gossip Communication: cannot reach clusters scus(ArgumentNullException)
```

I get this error on both clusters and only the cluster id is different.

Environment
Each cluster is an azure vm running 1 node service fabric. Each vm is in a different region but, connected by virtual network gateway. I can ping each vm. Additionally, i can telnet to the Orleans gateway port on both sides. I dont believe network connectivity is an issue in this case. I suspect i am missing something in the configuration or there is a defect with multi cluster when running on service fabric.

Each cluster has its own...
- storage provider connected to a region specific azure table
- log consistency provider connected to the storage provider above

Each cluster shares
- gossip channels on azure table storage. Connected to each of the region specific table storage accounts listed above

My asf stateless service hosting orleans. All the configuration is in code (nothing in app.config).
```csharp
public class OrleansStatelessService : StatelessService
{
    protected readonly IKernel _kernel;

    public OrleansStatelessService(StatelessServiceContext serviceContext, IKernel kernel)
        : base(serviceContext)
    {
        _kernel = kernel;
    }

    protected virtual ClusterConfiguration GetClusterConfiguration()
    {
        var config = new ClusterConfiguration();
        config.Globals.ReminderServiceType = GlobalConfiguration.ReminderServiceProviderType.ReminderTableGrain;
        //todo: set ClusterID to the name of the cluster on which this is deployed
        //todo: add gossipchannels
        config.Globals.DefaultMultiCluster = new List<string> { "scus", "eus" };
        config.Globals.ClusterId = "eus";
        config.Globals.GossipChannels = new List<GossipChannelConfiguration> {
            new GossipChannelConfiguration {
                ChannelType = GossipChannelType.AzureTable,
                ConnectionString = "{eus table storage account}"
            },
            new GossipChannelConfiguration {
                ChannelType = GossipChannelType.AzureTable,
                ConnectionString = "{scus table storage account}"
            }
        };
        config.Defaults.HostNameOrIPAddress = "10.0.0.4";
        config.Globals.MaxMultiClusterGateways = 10;
        config.Globals.DataConnectionString = "{eus table storage account}";
        config.UseStartupType<ClusterStartup>();
        config.Globals.UseGlobalSingleInstanceByDefault = true;
        LogManager.LogConsumers.Add(new EventSourceLogger());
        return config;
    }

    protected override IEnumerable<ServiceInstanceListener> CreateServiceInstanceListeners()
    {
        ClusterStartup.Service = this;
        ClusterStartup.Kernel = _kernel;
        return new[] { OrleansServiceListener.CreateStateless(this.GetClusterConfiguration()) };
    }

    protected override async Task RunAsync(CancellationToken cancellationToken)
    {
        try
        {
            ActorFactory _actors = new ActorFactory();
            var clusterlist = "scus,eus".Split(',');
            var mgtGrain = await _actors.Get<IManagementGrain>(0, this.Context.ServiceName);
            await mgtGrain.InjectMultiClusterConfiguration(clusterlist, "---EUS INJECTING CLUSTER CONFIG----");
        }
        catch (Exception ex)
        {

        }

        while (true)
        {
            cancellationToken.ThrowIfCancellationRequested();
            await Task.Delay(TimeSpan.FromSeconds(10), cancellationToken);
        }
    }

    protected override Task OnCloseAsync(CancellationToken cancellationToken)
    {
        if (ClusterStartup.Kernel != null)
            ClusterStartup.Kernel.Dispose();
        return base.OnCloseAsync(cancellationToken);
    }
}
```

subclass of stateless service above for the actual silo
```csharp
public class EngagementsService : OrleansStatelessService
{
    public const string LogConsistencyProviderName = "Default";
    public const string StorageProviderName = "Default";

    public EngagementsService(StatelessServiceContext serviceContext, IKernel kernel) : base(serviceContext, kernel)
    {
    }

    protected override ClusterConfiguration GetClusterConfiguration()
    {
        var config = base.GetClusterConfiguration();

        config.AddStateStorageBasedLogConsistencyProvider(LogConsistencyProviderName);
        config.Globals.RegisterDocumentDBProvider(StorageProviderName, "{eus doc db}", "{eus key}", "orleans", "grainstate");
        config.Globals.RegisterDocumentDBProvider("PubSubStore", "{eus doc db}", "{eus key}", "orleans", "grainstate");
        config.AddSimpleMessageStreamProvider(StreamProviders.StateChanges.Name, true);
        return config;
    }
}
```

**Comment by galvesribeiro** [comment]
cc: @sebastianburckhardt

**Comment by charleszipp** [comment]
I also updated to 1.5.0-rc binaries and still seeing the same issue.

The only record i see getting written is in the gossip table. I dont see anything else being written to grain state or any other tables being created

![image](https://user-images.githubusercontent.com/4292559/27542688-a4b480ea-5a4d-11e7-842b-db4c1f7dfa64.png)
![image](https://user-images.githubusercontent.com/4292559/27542697-ac00afc2-5a4d-11e7-8f57-c06b8c4aa4f3.png)

**Comment by sebastianburckhardt** [comment]
Thanks for sharing all the details right away! Very helpful.

All of your configuration looks right to me. I would probably remove the DefaultMultiCluster setting - no default is needed since your code explicitly injects the configuration - but this is not a problem either way.

The exception printed in the warning in the log indicates that the gossip mechanism cannot find an active multi-cluster gateway in the other cluster to talk to (clearly we need to improve the error message - it should not be a mysterious ArgumentNullException).

Now, why that is, I am not sure. 

What are the contents of the OrleansGossipTable? Do you see at least one gateway address for each cluster listed in there?

**Comment by charleszipp** [comment]
@sebastianburckhardt thanks for taking a look. Glad the info was helpful.

The record i showed in the screenshot above is the only record that exists in the gossip table. There are no gateway address' listed. If there should be, that makes more sense as i was wondering how each cluster would be able to determine the gateway if all thats in the gossip table is the cluster id. 

I have checked both gossip tables and what you see in the screen above is everything thats there. If you need the source to troubleshoot, i can send it directly but, probably cannot post it here. Its about 11MB with all projects and packages cleaned/removed.

**Comment by charleszipp** [comment]
So theres no ambiguity. Heres a screenshot of the table list and data together in one screen.

![image](https://user-images.githubusercontent.com/4292559/27547898-f1a7701a-5a5c-11e7-87e8-4310172bd26b.png)

**Comment by sebastianburckhardt** [comment]
Sorry, I overlooked the gossip table content that was already there in your first post! (it shows up very faint on my screen).

So, it looks like the clusters are not advertising their gateways - that's odd. And the multicluster configuration was indeed injected successfully. Not sure what's going on here. 

Are there any other warning or info messages by "MultiClusterOracle" in the trace file?

**Comment by sebastianburckhardt** [comment]
I just noticed that the ServiceId is 0 instead of a Guid. You may want to set it to some nonzero guid that identifies your service (must be the same on all clusters that are participating)  by setting `config.Globals.ServiceId`. 

But I don't think this will fix this problem. There must be some other reason why the gateways do not appear in the gossip channels.

**Comment by charleszipp** [comment]
@sebastianburckhardt the service id should be the same value on all clusters in the multicluster network right?

**Comment by charleszipp** [comment]
@sebastianburckhardt 

I tried assigning a non-zero guid as the service id to both clusters (both have the same id). Still getting same issue. I have attached the entire trace for each cluster (account keys redacted). The only other MultiClusterOracle warning is subsequent errors saying it cant communicate with the cluster. I looked at source code and this is basically just re-reporting the same error thats thrown at first. 

Ill dig through the source code some more to try to find out why it may not be writing the silo addresses while i wait on response.

[trace-eus.txt](https://github.com/dotnet/orleans/files/1103693/trace-eus.txt)
[trace-scus.txt](https://github.com/dotnet/orleans/files/1103694/trace-scus.txt)

Gossip Table contents is still the same

![image](https://user-images.githubusercontent.com/4292559/27563612-4e491c16-5a98-11e7-91d1-3d450e701898.png)

**Comment by sebastianburckhardt** [comment]
I suspect the Orleans cluster membership is not working as expected. This explains why there are no multi-cluster gateways, because the multi-cluster gateway list depends on the cluster membership component (called the 'MultiClusterOracle') for determining the gateway addresses.

There are several indications supporting this suspicion:

1. Looking at the logs I see no messages indicating that the MembershipOracle is starting, which is odd - when the silo is starting up it is supposed to print either "MembershipOracle starting" or "MembershipFailedToStart" - but I see neither one.

2. You say 
> I dont see anything else being written to grain state or any other tables being created

But there is supposed to be a table OrleansSiloInstances  listing the active silos... once cluster is started it should list the members of each cluster with status 'Active' and an IP address.

I am not quite sure why there is no membership oracle and no error message about it.

Actually, I just noticed that the configuration printed in the trace file shows
 `LivenessType: NotSpecified`
I am pretty sure this is not supposed to be unspecified. Maybe this could fix the problem:

```csharp
config.Globals.LivenessType = GlobalConfiguration.LivenessProviderType.AzureTable;
```

**Comment by charleszipp** [comment]
@sebastianburckhardt This doesnt even seem to be necessarily a multi-cluster issue anymore so much as its a service fabric integration issue if im reading what you are saying correctly. I ran a local instance of this on my laptop using azure table storage for everything and pointing it at azure storage emulator (i.e. UseDevelopmentStorage=true). Even in this case I dont see the OrleansSiloInstances table getting created despite adding the livelenesstype as AzureTable. I still do not see MembershipOracle referenced anywhere in the trace.

I had not used my local azure storage in a long while but, i noticed before i ran this test that the table you mentioned (OrleansSiloInstances) was sitting there. I think this was left over from when I first started testing orleans. At this time i was using 1.4.0. However, i eventually updated to 1.5.0 to get the instance version of the GrainClient. Im thinking this issue may be specific to 1.5. Im going to try to setup an example with latest 1.4 bits to see if i can reproduce. 

The only other thing i can think to do to help figure this out is attach my project to the orleans source to debug this. This will take a considerable amount of time from my end but, im not having much luck combing over the source code to find more ideas.

Should we loop anyone else in to get more eyes on this? This would seem to be a pretty serious 1.5 defect as it means no multi-cluster in 1.5 and potentially other negative impact from not having the OrleansSiloInstances table created.

**Comment by sebastianburckhardt** [comment]
My best guess right now is that your `ClusterStartup` class somehow overrides the membership provider, perhaps to use a SF-specific way of membership management.

Can you check ClusterStartup to see if it somehow sets the membership service to something nonstandard?

**Comment by sebastianburckhardt** [comment]
I think I understand now... I think for SF situations the silos are using a special class `FabricMembershipOracle` rather than the normal membership oracle. 

I think something is not working right in `FabricMembershipOracle`... unfortunately it does not trace anything at all (neither info, warning, or verbose messages) so it is hard to tell.

Perhaps @ReubenBond has an idea? He appears to be the author of the `FabricMembershipOracle`.

**Comment by charleszipp** [comment]
@sebastianburckhardt i created a new project from scratch with bare minimum and now im seeing the OrleansSiloInstances created wtih data about the location of the silo. I also am seeing data in the gossip table i didnt see in my other solution. i havent tested this on my multicluster vms yet but, dont think i need to do that to narrow down the root cause

I think this puts the problem back in something with my solution. Ill go through and slowly add back the components into this base solution to see if i can back into what is causing the issue. 

I had suspected my cluster startup as well. I was using this wire up ninject. Adding the source for what i was doing there. Basically trying to bridge ninject and orleans DI. I think i should be able to confirm if this was the issue or not by adding it to my now working barebones solution.

All the support is very much appreciated @sebastianburckhardt! 

```csharp
public static class ServiceCollectionContainerExtensions
{
    public static IServiceProvider BuildNinjectServiceProvider(this IServiceCollection services, IKernel kernel)
    {
        var rvalue = new NinjectServiceProvider(kernel);
        kernel.Bind<IServiceProvider>().ToConstant(rvalue);
        foreach (var service in services)
        {
            IBindingToSyntax<object> bind = kernel.Rebind(service.ServiceType);
            IBindingWhenInNamedWithOrOnSyntax<object> binding = null;
            if (service.ImplementationInstance != null)
                binding = bind.ToConstant(service.ImplementationInstance);
            else if (service.ImplementationType != null)
                binding = bind.To(service.ImplementationType);
            else
                binding = bind.ToMethod(ctx => service.ImplementationFactory(ctx.Kernel.Get<IServiceProvider>()));

            switch (service.Lifetime)
            {
                case ServiceLifetime.Scoped:
                    binding.InThreadScope();
                    break;

                case ServiceLifetime.Singleton:
                    binding.InSingletonScope();
                    break;

                case ServiceLifetime.Transient:
                    binding.InTransientScope();
                    break;
            }
        }

        return new NinjectServiceProvider(kernel);
    }
}

public class ClusterStartup
{
    public static StatelessService Service { get; set; }

    public static IKernel Kernel { get; set; }

    public IServiceProvider ConfigureServices(IServiceCollection services)
    {
        //registers an instance of the service fabric silo resolver using the provided service
        services.AddServiceFabricSupport(Service);

        //builds the default service provider. This could instead return an implentation that uses ninject.
        return services.BuildNinjectServiceProvider(Kernel);
    }
}

public class NinjectServiceProvider : IServiceProvider
{
    protected readonly IKernel _kernel;

    public NinjectServiceProvider(IKernel kernel)
    {
        _kernel = kernel;
    }

    public object GetService(Type serviceType)
    {
        return _kernel.TryGet(serviceType);
    }
}
```

**Comment by charleszipp** [comment]
@sebastianburckhardt @ReubenBond 

The thing that seems to be stopping those tables from being created is the 
```csharp
services.AddServiceFabricSupport(Service);
```
As soon as I plug that in, the OrleansSiloInstances table is not created. Additionally, the gossip table entries that contain the ip addresses of each cluster no longer get created.

Is this expected or a defect? Test solution that repro's the issue attached... 
[MultiClusterTest.zip](https://github.com/dotnet/orleans/files/1105862/MultiClusterTest.zip)

**Comment by sebastianburckhardt** [comment]
Yes, I understand now that SF does not create those tables at all but uses a custom membership provider class called `FabricMembershipOracle`. That's why we couldn't find anything in the log and no tables created. `FabricMembershipOracle` does not print anything to the log nor create any tables, so it is difficult to diagnose. What we believe to be happening is that `FabricMembershipOracle` is not returning anything to the multi-cluster logic, i.e. the call `GetApproximateMultiClusterGateways` is returning an empty list.

@ReubenBond may have some insight into how to inspect and/or troubleshoot the `FabricMembershipOracle` operation.

**Comment by charleszipp** [comment]
@ReubenBond or @sebastianburckhardt  Is this a bug in FabricMembershipOracle then? If i don't use this, can i still host the nodes on service fabric? Im testing and so far seems to work for local 1 node. Just have to setup the client to use the azure table gateway. I'll keep investigating this option as a workaround on multi nodes and multi-cluster while i wait on @ReubenBond's comments/suggestions.

**Comment by ReubenBond** [comment]
We are testing that `GetApproximateMultiClusterGateways()` is returning values according to the `MaxMultiClusterGateways` config value in the Fabric tests, however I haven't tested an actual multi-cluster setup yet.

I'm adding additional logging to `FabricMembershipOracle` which includes a snapshot of the multi-cluster gateways when they're queried. If it's too chatty, then we can change the severity from info to verbose.

**Comment by charleszipp** [comment]
@ReubenBond Thanks for looking into this. If it helps you, i could provide temporary access to the azure vm's i was using test this. They are already setup with visual studio, service fabric, and all the azure network resources needed to test cross-region multi-cluster network. If there's anything else i can provide to help let me know.

Is #3160 in the latest nuget packages? Is there a tag or milestone that would indicate that so i can monitor on my own?

**Comment by ReubenBond** [comment]
@charleszipp #3160 is not in `1.5.0-beta1` in the latest packages, but it will be included in `1.5.0` final and if there's a beta2, then it will be included in that.

If you're able, could you try testing with locally-built packages? To do that, open a terminal and run `build.cmd` from the root Orleans directory (master branch). That will output packages into the `Binaries\NuGet.Packages\Prerelease` folder. You can add that as a NuGet package source in VS or by creating a file called `NuGet.Config` in your solution directory with contents like this:
```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="orleans-local" value="file://c:/dev/orleans/Binaries/NuGet.Packages/Prerelease" />
  </packageSources>
</configuration>
```

Then upgrade to those packages, which will be suffixed with -Dev + the build time.
![image](https://user-images.githubusercontent.com/203839/27607204-7158c542-5bc6-11e7-9873-07725a3b2f72.png)

**Comment by charleszipp** [comment]
Thanks @ReubenBond. I will give this a shot and report back here my findings.

**Comment by charleszipp** [comment]
@ReubenBond wanted to provide a quick update. So far things are looking good. I have only tested this locally on my laptop first before porting it over to my multi-cluster environment vms. I am noticing the gossip table does have the cluster IP and status. It did not have this before. I also am noticing a few new tables, OrleansSiloMetrics and OrleansSiloStatistics. See screens below

Next im going to try to move this over to the vm's and verify the two clusters are able to communicate.

![image](https://user-images.githubusercontent.com/4292559/27637130-33a171dc-5bd4-11e7-86e7-3c792a1f12fc.png)

![image](https://user-images.githubusercontent.com/4292559/27637134-38fec184-5bd4-11e7-8a2f-537164b3ceed.png)

cc @sebastianburckhardt

**Comment by charleszipp** [comment]
@ReubenBond confirmed cross region multi-cluster is now working for global single instance grains using latest from master. I had two instances of my cluster in different vms with visual studio debugging attached. I activated a grain on eus and confirmed state was stored in the eus state provider. I then tried to activate the same grain from scus cluster. My breakpoint for eus cluster's grain was hit! It successfully routed the request to the appropriate cluster.

One question i had is one what is the difference between DeploymentId and ServiceId. I had set the deployment id this time as it was needed when i was testing without using the FabricMembershipOracle. Can i set these to have the same guid?

Next step is to try one instance per cluster and see how that behaves. However, i think its safe to conclude this definitely works in latest master. Whether or not it was fixed with #3160 i think depends on making sure not providing the deployment id wasnt the issue. Ill test that and after i think we should be able to close this as either fixed with #3160 or configuration issue.

**Comment by charleszipp** [comment]
@ReubenBond removed deployment id and still appears to be working. I think we can close this as it appears to be fixed in latest master just not working in 1.5.0-rc.

Thanks for the assistance @ReubenBond  and @sebastianburckhardt. Much appreciated. Looking forward to 1.5.0 GA release!
